target_chains/cosmwasm/tools/src/pipeline.ts

```diff
@@ -0,0 +1,216 @@
+  }
+
+  // After all the in memory operations one can commit to the local file
+  // for permanent storage
+  commit() {
+    writeFileSync(this.filePath, JSON.stringify(this.store, null, 4));
+  }
+}
+
+// A step is defined as a method
+// it takes in a step id as a parameter to execute that particular step
+// Optionally it can take in a method `getResultOfPastStage` as a parameter
+// if it wants to access the result of the previous stages
+export type Step =
+  | ((
+      stepId: string,
+      // get the result of a past stage using it's id
+      // It will return the result for the same step id
+      // It will return undefined if the previous stage data has not been stored locally
+      // or if a future stage data is being asked
+      getResultOfPastStage: <Y>(stageId: string) => Y
+    ) => Promise<any>)
+  | ((stepId: string) => Promise<any>);
+
+// A step can fail. If the error is not handled it will crash the pipeline
+// We would like to store the result or reason locally too
+// For that purpose the response from a step is being wrapped in this
+type StepResult<T = any> =
+  | {
+      status: "rejected";
+      stepId: string;
+      reason: any;
+    }
+  | {
+      status: "fulfilled";
+      stepId: string;
+      result: T;
+    };
+
+// A stage will contain an identifier and a step method
+type Stage = {
+  stageId: string;
+  step: Step;
+};
+
+// A pipeline is conists of multiple stages
+// A stage contains of multiple steps
+// Stages will be run synchronously i.e, in order
+// Steps will be run asychronously
+// CONSTRAINT: Each stage will have the same number of steps
+// See the type definition of `Stage` and `Step` above to know more about them
+export class Pipeline {
+  private readonly stages: Stage[] = [];
+  constructor(
+    private readonly version: string,
+    private readonly stepIds: string[]
+  ) {}
+
+  addStage(stageId: string, step: Step) {
+    this.stages.push({ stageId, step });
+  }
+
+  // We want to wrap the step provided by the pipeline consumer
+  // In order to wrap the response of the step in the StepResult
+  // also in this method we inject the `getResultOfPastStage` to the step
+  private stepWrapper<T, Y>(step: Step) {
+    return async (stepId: string): Promise<StepResult<T>> => {
+      // method to inject
+      const getResultOfPastStage = <Y>(stageId: string): Y => {
+        let stateManager = new StateManager<StepResult<Y>>(
+          stageId,
+          this.version
+        );
+        let result = stateManager.getStepState(stepId);
+        // pipeline will only proceed to the next stage if the previous one is fulfilled
+        if (result !== undefined && result.status === "rejected") {
+          throw new Error("previous stage was not fulfilled");
+        }
+        if (result === undefined) {
+          throw new Error(
+            `either the previous stage is not fully processed or a future stage is being referred to: ${stageId}`
+          );
+        }
```

Comment:
why is this error logic in here? seems like if a stage failed, the pipeline shouldn't even be trying to run the next stage?